lib/sync/replication-state.js

```diff
@@ -0,0 +1,279 @@
+import { truncate } from 'fs'
+import { TypedEmitter } from 'tiny-typed-emitter'
+import { idToKey, keyToId, truncateId } from '../utils.js'
+
+export class ReplicationState extends TypedEmitter {
+  #coreStates = new Map()
+
+  /**
+   * @param {Object} options
+   * @param {import('../core-manager/index.js').Namespace} options.namespace
+   * @param {import('../core-manager/index.js').CoreManager} options.coreManager
+   */
+  constructor(options) {
+    super()
+    this.coreManager = options.coreManager
+    this.namespace = options.namespace
+
+    const coreRecords = this.coreManager.getCores(this.namespace)
+    for (const coreRecord of coreRecords) {
+      this.#core(coreRecord)
+    }
+
+    this.coreManager.on('add-core', (coreRecord) => {
+      this.#core(coreRecord)
+    })
+  }
+
+  get state () {
+    const cores = [...this.#coreStates.entries()]
+      .reduce((obj, [ corePublicId, state ]) => {
+        obj[corePublicId] = state.state
+        return obj
+      }, {})
+
+    const synced = this.isSynced()
+
+    return { cores, synced }
+  }
+
+  get peers () {
+    const peers = new Set()
+
+    for (const state of this.#coreStates.values()) {
+        for (const peer of state.peers) {
+            peers.add(peer)
+        }
+    }
+
+    return [...peers]
+  }
+
+  isSynced () {
+    for (const state of this.#coreStates.values()) {
+        if (!state.isSynced()) {
+            return false
+        }
+    }
+
+    return true
+  }
+
+  close() {
+    this.removeAllListeners()
```

Comment:
So far this class only adds listeners to itself, consumers don't add listeners. I imagined that this class would be internal to the ReplicationState class that would surface state of all cores to a consumer. If that ends up not being the case then I think this can be revisited.